Public/Src/Cache/ContentStore/Distributed/NuCache/ReplicatedRedisHashKey.cs

```diff
@@ -0,0 +1,210 @@
+            }
+
+            if (!primaryVersionedResult.Succeeded)
+            {
+                // Error in primary, just use secondary
+                return new Result<T>(secondaryVersionedResult.Value.result, isNullAllowed: true);
+            }
+
+            // Prefer db with highest version number (or primary if equal)
+            bool preferPrimary = primaryVersionedResult.Value.version >= secondaryVersionedResult.Value.version;
+
+            if (_host.CanMirror && !_lastMirrorTime.IsRecent(_clock.UtcNow, _host.MirrorInterval))
+            {
+                _lastMirrorTime = _clock.UtcNow;
+
+                if (preferPrimary)
+                {
+                    // Primary is preferred, try to backup to secondary
+                    await TryMirrorRedisHashDataAsync(
+                        context,
+                        source: _redis.PrimaryRedisDb,
+                        target: _redis.SecondaryRedisDb);
+                }
+                else
+                {
+                    await TryMirrorRedisHashDataAsync(
+                        context,
+                        source: _redis.SecondaryRedisDb,
+                        target: _redis.PrimaryRedisDb,
+
+                        // Set version of secondary db to its prior version so primary will now take precedence
+                        // Primary will have the current version of secondary after mirroring and secondary will have 
+                        postMirrorSourceVersion: secondaryVersionedResult.Value.version);
+                }
+            }
+
+            var result = preferPrimary ? primaryVersionedResult.Value.result : secondaryVersionedResult.Value.result;
+            return new Result<T>(result, isNullAllowed: true);
+        }
+
+        private async Task TryMirrorRedisHashDataAsync(OperationContext context, RedisDatabaseAdapter source, RedisDatabaseAdapter target, long? postMirrorSourceVersion = null)
+        {
+            await context.PerformOperationAsync(
+                _host.Tracer,
+                async () =>
+                {
+                    var sourceDump = await source.ExecuteBatchAsync(context,
+                        b => b.AddOperation(_key, b => b.KeyDumpAsync(_key)),
+                        RedisOperation.HashGetKeys);
+
+                    await target.ExecuteBatchAsync(context,
+                        b =>
+                        {
+                            var deleteTask = b.AddOperation(_key, b => b.KeyDeleteAsync(_key));
+                            var restoreTask = b.AddOperation(_key, b => b.KeyRestoreAsync(_key, sourceDump).WithResultAsync(Unit.Void));
+                            return Task.WhenAll(deleteTask, restoreTask).WithResultAsync(Unit.Void);
+                        },
+                        RedisOperation.HashDeleteAndRestore);
+
+                    if (postMirrorSourceVersion.HasValue)

```

Comment:
Just curious: should the version update be par tof the second execute batch to make the update "atomic"?

Reply:
The second update batch is to the target. This is to the source. We only want to modify the version if the target update succeeded.

---
In reply to: [337752360](https://github.com/microsoft/BuildXL/pull/1081#discussion_r337752360) [](ancestors = 337752360)